File: src/modules/graphs/2_Traversals/DFS.ts
class Graph {
    public vertices: number;
    public adjacencyList: Array<Array<number>>;
    constructor(vertices: number) {
        this.vertices = vertices;
        this.adjacencyList = new Array(vertices).fill(null).map(() => []);
    }

    addEdge(source: number, destination: number) {
        this.adjacencyList[source].push(destination);
        this.adjacencyList[destination].push(source); // For undirected graphs
    }

    dfs(start: number) {
        const visited = new Array(this.vertices).fill(false);
        const stack = [];
        stack.push(start);

        while (stack.length > 0) {
            const vertex = stack.pop();
            if (vertex === undefined) break;
            if (!visited[vertex]) {
                process.stdout.write(vertex + ' ');
                visited[vertex] = true;

                for (const neighbor of this.adjacencyList[vertex]) {
                    if (!visited[neighbor]) {
                        stack.push(neighbor);
                    }
                }
            }
        }
    }
}

const vertices = 7;
const graph = new Graph(vertices);

graph.addEdge(0, 1);
graph.addEdge(0, 2);
graph.addEdge(1, 3);
graph.addEdge(1, 4);
graph.addEdge(2, 5);
graph.addEdge(2, 6);

console.log("Depth-First Traversal (starting from vertex 0):");
graph.dfs(0);

export { }

File: src/modules/graphs/2_Traversals/DFS.go
package main

import "fmt"

type Graph struct {
	vertices      int
	adjacencyList [][]int
}

func newGraph(vertices int) *Graph {
	list := make([][]int, vertices)
	for i := range list {
		list[i] = make([]int, 0)
	}

	return &Graph{vertices: vertices, adjacencyList: list}
}

func (g *Graph) AddEdge(source, destination int) {
	g.adjacencyList[source] = append(g.adjacencyList[source], destination)
	g.adjacencyList[destination] = append(g.adjacencyList[destination], source) // as we have undirected graph
}

func (g *Graph) Dfs(start int) []int {
	visited := make([]bool, g.vertices)
	stack := []int{}

	output := []int{}

	stack = append(stack, start)

	for len(stack) > 0 {
		curr := stack[len(stack)-1]
		stack = stack[:len(stack)-1]

		if !visited[curr] {
			output = append(output, curr)
			visited[curr] = true

			for _, neighbor := range g.adjacencyList[curr] {
				if !visited[neighbor] {
					stack = append(stack, neighbor)
				}
			}
		}
	}

	return output
}

func main() {
	graph := newGraph(7)

	graph.AddEdge(0, 1)
	graph.AddEdge(0, 2)
	graph.AddEdge(1, 3)
	graph.AddEdge(1, 4)
	graph.AddEdge(2, 5)
	graph.AddEdge(2, 6)

	dfsTraversal := graph.Dfs(0)
	fmt.Println(dfsTraversal)
}
